plugin-server/src/worker/ingestion/persons/batch-writing-person-store.ts

```diff
@@ -961,4 +923,101 @@ export class BatchWritingPersonsStoreForBatch implements PersonsStoreForBatch, B
+                        if (refreshedPersonUpdate) {
+                            personUpdate = refreshedPersonUpdate
+                            continue
+                        }
+                        // If we can't refresh the person ID, we can't retry, fail gracefully
+                        return [toInternalPerson(personUpdate), []]
+                    }
+
+                    // Handle optimistic update conflicts with special logging
+                    if (error instanceof RaceConditionError) {
+                        logger.warn(`Optimistic update conflict for ${operation}, retrying...`, {
+                            attempt,
+                            maxRetries,
+                            teamId: personUpdate.team_id,
+                            personId: personUpdate.id,
+                            distinctId: personUpdate.distinct_id,
+                        })
+                    } else {
+                        // For any other error type, still retry but with generic logging
+                        logger.warn(`Database error for ${operation}, retrying...`, {
+                            attempt,
+                            maxRetries,
+                            teamId: personUpdate.team_id,
+                            personId: personUpdate.id,
+                            distinctId: personUpdate.distinct_id,
+                            error: error instanceof Error ? error.message : String(error),
+                        })
+                    }
+
+                    await new Promise((resolve) => setTimeout(resolve, retryInterval))
+                    continue
+                }
+
+                throw error
+            }
+        }
+
+        // This should never be reached, but TypeScript requires it
+        throw new Error('Unexpected end of retry loop')
+    }
+
+    /**
+     * Refreshes the person ID for a given distinct ID by fetching from the database.
+     * This handles cases where the person was merged and the ID changed.
+     * @param personUpdate the PersonUpdate that failed to update
+     * @returns updated PersonUpdate with new person ID if found, null if person no longer exists
+     */
+    private async refreshPersonIdAfterMerge(personUpdate: PersonUpdate): Promise<PersonUpdate | null> {
+        const currentPerson = await this.db.fetchPerson(personUpdate.team_id, personUpdate.distinct_id)
+
+        if (!currentPerson) {
+            // Person truly doesn't exist anymore
+            return null
+        }
+
+        // Clear the old person ID from cache since it's been merged
+        this.clearPersonCacheForPersonId(personUpdate.team_id, personUpdate.id)
+
+        // Update our cache mapping to reflect the new person ID
+        this.setDistinctIdToPersonId(personUpdate.team_id, personUpdate.distinct_id, currentPerson.id)
+
+        // Create updated PersonUpdate with the new person ID and version
+        const updatedPersonUpdate = {
+            ...personUpdate,
```

Comment:
thought: I feel that one day we're gonna add something to person updates and forget to update this. WDYT about making the fields explicit here? It's a bit more code, but if the interface changes it will be caught by the compiler.